docs/proposals/20221003-In-place-propagation-of-Kubernetes-objects-only-changes.md

```diff
@@ -0,0 +1,321 @@
+
+### Propagation of fields impacting only Kubernetes objects or controller behaviour
+
+In addition to labels and annotations, there are also other fields that flow down from ClusterClass to KubeadmControlPlane/MachineDeployments and
+ultimately to Machines.
+
+Some of them can be considered like labels and annotations, because they have impacts only on Kubernetes objects or controller behaviour, but
+not on the actual Machine itself - including infrastructure and the software running on it  (in-place mutable fields).
+Examples are `MinReadySeconds`, `NodeDrainTimeout`, `NodeVolumeDetachTimeout`, `NodeDeletionTimeout`.
+
+Propagation of changes to those fields will be implemented using the same [in-place propagation](#in-place-propagation) mechanism implemented
+for metadata.
+
+### In-place propagation
+
+With in-place propagation we are referring to a mechanism that updates existing Kubernetes objects, like MachineSets or Machines, instead of
+creating a new object with the updated fields and then deleting the current Kubernetes object.
+
+The main benefit of this approach is that it prevents unnecessary rollouts of the corresponding infrastructure, with the consequent creation/
+deletion of a Kubernetes node and drain/scheduling of workloads hosted on the Machine being deleted.
+
+**Important!** In-place propagation of changes as defined above applies only to metadata changes or to fields impacting only Kubernetes objects
+or controller behaviour. This approach can not be used to apply changes to the infrastructure hosting a Machine, to the OS or any software
+installed on it, Kubernetes components included (Kubelet, static pods, CRI etc.).
+
+Implementing in-place propagation has two distinct challenges:
+
+- Current rules defining when MachineDeployments or KubeadmControlPlane trigger a rollout should be modified in order to ignore metadata and
+  other fields that are going to be propagated in-place.
+
+- When implementing the reconcile loop that performs in-place propagation, it is required to avoid impact on other components applying
+  labels or annotations to the same object. For example, when reconciling labels to a Machine, Cluster API should take care of reconciling
+  only the labels it manages, without changing any label applied by the users/by another controller on the same Machine.
+
+#### MachineDeployment rollouts
+
+The MachineDeployment controller determines when a rollout is required using a "semantic equality" comparison between current MachineDeployment
+spec and the corresponding MachineSet spec.
+
+While implementing this proposal we should change the definition of "semantic equality" in order to exclude metadata and fields that
+should be updated in-place.
+
+On top of that we should also account the use case where, after deploying the new "semantic equality" rule, there is already one or more
+MachineSet(s) matching the MachineDeployment. In this case Cluster API deterministically picks the oldest of them.
```

Comment:
Just to make sure I understood it correctly:

Today: `Cluster API deterministically picks the oldest of them.`

After this proposal is implemented: `pick the "semantically equal" MachineSet with more
Machines`?

If I"m correct maybe let's add "today" to the first statement to make it 100% clear that's the current behavior and not the proposed one.